File: quals/web/greycat-game/solution/solve.py
#!/usr/bin/env python3
import json
import sys
import time
import base64
import urllib.parse
import urllib.request
from http.cookiejar import CookieJar


BASE_URL = (sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:34467").rstrip("/")


def make_opener():
    return urllib.request.build_opener(urllib.request.HTTPCookieProcessor(CookieJar()))


def get_json(opener, path, headers=None):
    req = urllib.request.Request(f"{BASE_URL}{path}", headers=headers or {})
    with opener.open(req) as resp:
        return json.loads(resp.read().decode()), resp.info()


def decode_stamp(stamp, trace_id):
    encoded = base64.b64decode(stamp)
    parts = str(trace_id or "").split("-")
    seed = parts[1] if len(parts) >= 3 else ""
    index = int(parts[2]) - 1 if len(parts) >= 3 and parts[2].isdigit() else -1
    key_base = sum(ord(ch) for ch in seed) + max(0, index) * 17
    decoded = bytes(
        value ^ ((key_base + offset * 13) & 0xFF)
        for offset, value in enumerate(encoded)
    )
    return decoded.decode("utf-8")


def main():
    opener = make_opener()
    bootstrap, _ = get_json(opener, "/api/bootstrap")
    target = bootstrap["fastPhaseScore"] + 40

    samples = [
        (0, 0),
        (120, 260),
        (240, 620),
        (360, 1180),
        (480, 1880),
        (560, target),
    ]

    for tick, score in samples:
        query = urllib.parse.urlencode({"tick": tick, "score": score, "state": "running"})
        get_json(opener, f"/api/run?{query}")
        time.sleep(1.1)

    fragments = []
    for lane in (0, 1, 0, 1, 0, 1):
        query = urllib.parse.urlencode({"score": target, "lane": lane})
        ghost, _ = get_json(opener, f"/api/ghost?{query}")
        stamp = ghost.get("stamp")
        trace_id = ghost.get("traceId")
        if stamp and trace_id:
            fragments.append(decode_stamp(stamp, trace_id))

    flag = "".join(fragments)
    print(flag)


if __name__ == "__main__":
    main()

File: quals/web/greycat-game/solution/solve.js
const baseUrl = (process.argv[2] || "http://127.0.0.1:34467").replace(/\/+$/, "");

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function extractCookie(headers) {
  const raw = headers.get("set-cookie");
  if (!raw) {
    return "";
  }
  return raw.split(";")[0];
}

function decodeStamp(stamp, traceId) {
  const encoded = Buffer.from(stamp, "base64");
  const parts = String(traceId || "").split("-");
  const seed = parts.length >= 3 ? parts[1] : "";
  const index = Number(parts.length >= 3 ? parts[2] : 0) - 1;
  const keyBase =
    seed.split("").reduce((sum, ch) => sum + ch.charCodeAt(0), 0) + Math.max(0, index) * 17;
  const bytes = [];

  for (let i = 0; i < encoded.length; i += 1) {
    bytes.push(encoded[i] ^ ((keyBase + i * 13) & 0xff));
  }

  return Buffer.from(bytes).toString("utf8");
}

async function requestJson(url, options = {}, cookie = "") {
  const headers = {
    ...(options.headers || {}),
  };
  if (cookie) {
    headers.Cookie = cookie;
  }

  const response = await fetch(url, {
    ...options,
    headers,
  });

  const text = await response.text();
  let json;
  try {
    json = JSON.parse(text);
  } catch (error) {
    throw new Error(`Non-JSON response from ${url}: ${text}`);
  }

  return {
    response,
    json,
    cookie: extractCookie(response.headers) || cookie,
  };
}

async function main() {
  let cookie = "";

  const bootstrap = await requestJson(`${baseUrl}/api/bootstrap`, {}, cookie);
  cookie = bootstrap.cookie;
  const fastPhaseScore = bootstrap.json.fastPhaseScore;

  const runSamples = [
    { tick: 0, score: 0 },
    { tick: 90, score: 180 },
    { tick: 180, score: 420 },
    { tick: 300, score: 980 },
    { tick: 420, score: 1640 },
    { tick: 500, score: 2010 },
    { tick: 540, score: fastPhaseScore + 40 },
  ];

  for (const sample of runSamples) {
    const run = await requestJson(
      `${baseUrl}/api/run?score=${sample.score}&tick=${sample.tick}&state=running`,
      {},
      cookie,
    );
    cookie = run.cookie;
    await sleep(1100);
  }

  const fragments = [];
  for (let attempt = 0; attempt < 12; attempt += 1) {
    const ghost = await requestJson(
      `${baseUrl}/api/ghost?score=${fastPhaseScore + 50}&lane=${attempt % 2}`,
      {},
      cookie,
    );
    cookie = ghost.cookie;
    if (ghost.json.stamp && ghost.json.traceId) {
      fragments.push(decodeStamp(ghost.json.stamp, ghost.json.traceId));
      if (fragments.join("").endsWith("}")) {
        break;
      }
    }
  }

  const flag = fragments.join("");

  console.log(`base_url=${baseUrl}`);
  console.log(`session=${bootstrap.json.session}`);
  console.log(`fast_phase_score=${fastPhaseScore}`);
  console.log(`fragments=${JSON.stringify(fragments)}`);
  console.log(`flag=${flag}`);
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});
